### ISSUE TITLE
Materialize and/or to_pandas_df in parallel.

### ISSUE THREAD

**Issue by melovv** [vaexio/vaex#259]
Hello, guys.

I just found Vaex and am trying to use it in a project. I'm more interested in the expression system. 

Right now, I have a table with tens of virtual columns and would like to materialize them in parallel and/or export them to a pandas dataframe in parallel. Is that possible? When I run df.to_pandas_df(virtual=True) it is using a single core.

Thanks a lot!

Vinícius.

**Comment by maartenbreddels** [comment]
Hi Vinicius,

that is a good observations, at the core of the problem is `DataFrame.evaluate` which is single threaded. In the early days this was the core entry point which all threads used to evaluate, hence it didn't multithread again. This is a nuisance for me as well, and high on my wishlist.
A few workarounds:
 * export to hdf5, only useful if too large to fit into memory and you plan to reuse it
 * use df.map_reduce, sth like this:
```python
import vaex
import numpy as np
df = vaex.open('/Users/maartenbreddels/datasets/nyc_taxi2015.arrow')
df['r'] = df.dropoff_latitude + 1
df['q'] = df.dropoff_longitude + 1
virtual_columns = ['r', 'q']

outputs = [np.zeros(len(df), dtype=df.dtype(k)) for k in virtual_columns]
                    
def map(thread_index, i1, i2, *chunks):
    for output, chunk in zip(outputs, chunks):
        output[i1:i2] = chunk
def reduce(a, b):
    pass
df.map_reduce(map, reduce, virtual_columns, info=True)
```
This ensures all chunks are done in parallel, with 1 pass over the data. The last point is important when the expressions depend a lot on the same columns.

Does this help?

I'll keep this issue open until we have a nice solution + documentation on how to do this properly.

Cheers,

Maarten

**Comment by melovv** [comment]
Hi, Maarten.

Thanks for the quick reply and the code. It runs super fast!
I'm gonna open another ticket regarding another topic.

Cheers,

Vinícius.

**Comment by maartenbreddels** [comment]
This should be solved in #474
